src/Castle.Core/DynamicProxy/ProxyGenerator.cs

```diff
@@ -1595,5 +1596,136 @@ protected List<object> BuildArgumentListForClassProxy(ProxyGenerationOptions opt
 			// create proxy
 			return ProxyBuilder.CreateClassProxyTypeWithTarget(classToProxy, additionalInterfacesToProxy, options);
 		}
+
+
+		/// <summary>
+		///   Creates a delegate that, when invoked, calls the specified <paramref name="interceptors"/>.
+		/// </summary>
+		/// <typeparam name="TDelegate">Type of delegate which will be proxied.</typeparam>
+		/// <param name="interceptors">The interceptors called during the invocation of the delegate.</param>
+		/// <returns>
+		///   New delegate of type <typeparamref name="TDelegate"/> that, when invoked, calls the specified <paramref name="interceptors"/>.
+		/// </returns>
+		/// <exception cref="ArgumentException">Thrown when the given <typeparamref name="TDelegate"/> is a generic type definition, or not a delegate type.</exception>
+		/// <remarks>
+		///   This method uses <see cref="IProxyBuilder.CreateDelegateProxyType(Type)"/> to generate a delegate.
+		///   As such, the caller should additionally expect any type of exception that the given <see cref="IProxyBuilder"/> implementation may throw.
+		/// </remarks>
+		public TDelegate CreateDelegateProxy<TDelegate>(params IInterceptor[] interceptors)
+		{
+			return (TDelegate)(object)CreateDelegateProxyImpl(typeof(TDelegate), null, interceptors);
+		}
+
+		/// <summary>
+		///   Creates a delegate that, when invoked, calls the specified <paramref name="interceptors"/>.
+		/// </summary>
+		/// <typeparam name="TDelegate">Type of delegate which will be proxied.</typeparam>
+		/// <param name="target">The target. Must have type <typeparamref name="TDelegate"/>.</param>
+		/// <param name="interceptors">The interceptors called during the invocation of the delegate.</param>
+		/// <returns>
+		///   New delegate of type <typeparamref name="TDelegate"/> that, when invoked, calls the specified <paramref name="interceptors"/>.
+		/// </returns>
+		/// <exception cref="ArgumentNullException">Thrown when <paramref name="target"/> is a null reference (Nothing in Visual Basic).</exception>
+		/// <exception cref="ArgumentException">Thrown when the given <typeparamref name="TDelegate"/> is a generic type definition, or not a delegate type.</exception>
+		/// <remarks>
+		///   This method uses <see cref="IProxyBuilder.CreateDelegateProxyType(Type)"/> to generate a delegate.
+		///   As such, the caller should additionally expect any type of exception that the given <see cref="IProxyBuilder"/> implementation may throw.
+		/// </remarks>
+		public TDelegate CreateDelegateProxyWithTarget<TDelegate>(TDelegate target, params IInterceptor[] interceptors)
+		{
+			if (target == null)
+			{
+				throw new ArgumentNullException(nameof(target));
+			}
+
+			return (TDelegate)(object)CreateDelegateProxyImpl(typeof(TDelegate), (Delegate)(object)target, interceptors);
```

Comment:
These ugly type casts could be avoided by adding a `where TDelegate : Delegate` constraint. This would require C# 7.3, which unfortunately Mono does not yet support, even in the current stable version (5.12).